wait_all.go

```diff
@@ -0,0 +1,37 @@
+package asynctask
+
+// WaitAll block current thread til all task finished.
+// it return immediately after receive first error.
+func WaitAll(tasks ...*TaskStatus) error {
+	errorCh := make(chan error)
+	errorChClosed := false
+	defer func() {
+		errorChClosed = true
+		close(errorCh)
+	}()
+
+	runningTasks := 0
+	for _, tsk := range tasks {
+
+		go func(tsk *TaskStatus) {
+			_, err := tsk.Wait()
+			if !errorChClosed {
+				errorCh <- err
+			}
+		}(tsk)
+		runningTasks++
+	}
+
+	for {
+		select {
+		case err := <-errorCh:
+			runningTasks--
+			if err != nil && isErrorReallyError(err) {
```

Comment:
why you want return on the first error?

Reply:
usually when one error occurs, you likely to abort the whole operation, right? then it make no sense to wait for the rest.

also there could be case where error can be handled and the operation still can do meaningful things.

It seems I should introduce an option or different func, so caller can decide what behavior it want.

Reply:
yes. I think there are cases you want to wait all and there are cases you want to bail early.